debug_module.tex

```diff
@@ -165,26 +165,34 @@ \section{Hart States}
 
 \section{Run Control} \label{runcontrol}
 
-For every hart, the Debug Module contains 4 conceptual bits of state: halt
-request, resume request, halt-on-reset request,  and hart reset.
-(The hart reset and halt-on-reset request bits are optional.)
-These bits all reset to 0. A debugger can write them for the currently selected
-harts through \Fhaltreq, \Fresumereq, \Fsetresethaltreq/\Fclrresethaltreq and
-\Fhartreset in \Rdmcontrol, only one of which may be high in any single write.
+For every hart, the Debug Module contains 3 conceptual bits of state: halt
+request, halt-on-reset request, and hart reset.  (The hart reset and
+halt-on-reset request bits are optional.) These bits all reset to 0. In
+addition there are resume request signal and clear resume ack signals.
+A debugger can write these bits/signals for the currently selected
+harts through \Fhaltreq, \Fresumereq, \Fsetresethaltreq/\Fclrresethaltreq, and
+\Fhartreset in \Rdmcontrol.
 In addition the DM receives halted, running, and resume ack signals from each
 hart.
 
-When a running hart receives a halt request, it responds by halting,
+% In my head bits imply flip-flops, while signals might just be the output of
+% combinational logic/state machine. Does that need more clarification?
+
+To halt a hart, the debugger writes 1 to \Fhaltreq, which sets the halt request
+bit for each selected hart. When a running hart sees a halt request, it responds
+by halting,
 deasserting its running signal, and asserting its halted signal. The halted
 signals of all selected harts are reflected in the \Fallhalted and \Fanyhalted
-bits. \Fhaltreq is ignored by halted harts.
-
-When a halted hart receives a resume request, it responds by resuming, clearing
-its halted signal, and asserting its running signal and resume ack signals. The
-resume ack signal is lowered when the resume request is deasserted.  These
+bits. Halted harts ignore the halt request bit.
+
+To resume a hart, the debugger writes 1 to \Fresumereq and 0 to \Fhaltreq. This
+causes the DM to send each selected hart a resume request.
+The selected, halted harts respond by resuming, clearing
+their halted signals, and asserting their running and resume ack signals. To
+clear resume ack, the debugger writes 0 to \Fresumereq, which results in the
+DM sending a clear resume ack signal to the hart.  These
```

Comment:
Was this always the functionality? Or was this changed more recently (that clearing resumereq clears resumeack).

Reply:
resumereq used to be a R/W bit, and it had been the behavior that when it was written to 0 that resumeack would go low. The new wording is almost identical to that, except it's (I think) more clear that writing resumeack 1 only causes resume to happen once. Also a DM doesn't have to store the resumereq value anymore if it doesn't want to. (Eg. in a complex system it could dispatch a message.)

Reply:
I don't see why we need to specify that writing resumereq to 0 causes resumeack to go low. You don't really care about resumeack until you write resumereq to 1. So I think that should be the specified behavior (pretty sure it used to be anyway).

Reply:
When do you think resumeack should go low? It has to happen at some point.

When resumereq is written is too late, because there's a race:
1. debugger writes resumereq
2. debugger checks resumeack, which is still high
3. because something was being really slow, only now does the hart see the resume request, clearing its resume ack

I think as long as the debugger can see resumeack go low before it performs a resumereq, that eliminates race conditions like that.

Reply:
We need to be able to write resumereq to 0 without having any effect, because we want to be able to .

I think that resumeack should go low once a new resumereq written to 1. No one cares what resumeack is until they write resumereq to 1 again.

This does assume that resumeack goes low instantly upon writing new resumereq, but this is similar to requirements that ABSTRACTCS.busy goes high immediately, or SBCS.busy goes high immediately.

Reply:
That sounds right. I've updated the PR.